Add the ability to obtain the number of valid validator at the next block height

MichaelKim20 opened the pull request:
This is the first PR to solve issue 822.
Minimum valid validators must exist at all block heights. This ensures that the following blocks are continuously created:
However, this should not be checked after the block is made. It is necessary to check this during the consensus process.
So we need to get the number of valid validators at the next block height, when the blocks that are now being agreed upon are applied.

I implemented this.


Relates to Issue #822

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -319,6 +320,35 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    public void getAvailableKeys (ulong block_height, ref Set!Hash utxo_key_set) @safe nothrow
```
The `utxo_key_set` seems unnecessary. This function could be renamed `ulong getValidatorCount (ulong block_height)`. Then all you have to do is call:

```D
this.db.execute("SELECT count(*) FROM validator_set WHERE enrolled_height >= ?", height).oneValue!ulong;
```

MichaelKim20 replied:
Shouldn't there be more than X validators?
If only one validator is needed, I agree with you. 
However, When we need several, shouldn't we remove the same one and check the number?

AndrejMitrovic replied:
`SELECT count(*)` returns the count.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -811,6 +812,28 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public ulong getAvailableCount (ulong block_height, ref Enrollment[] enrolls) @safe nothrow
```
`getValidatorCount` would be a better name, and it should just forward to the implementation in `ValidatorSet`.

You don't need to pass the enrollments here. Think about it, if a block will contain enrollments it means the validator set will not be empty.

There is only a single protocol check that we need to do:

- If the enrollments in the proposed `ConsensusData` are empty, then check the ValidatorSet and make sure that there will still be valid enrollments for the proposed new block height.

MichaelKim20 replied:
Shouldn't there be more than X validators?
If only one validator is needed, I agree with you.
However, When we need several, shouldn't we remove the same one and check the number?

AndrejMitrovic replied:
What I mean is, is that the calling function can already do this. It doesn't need to pass the `Enrollment[]` here. It could either pass the length, or it could calculate it at the call site.

So the code in Nominator could do something like this:

```D
if (consensus_data.enrollments.length + enrollment_man.getValidatorCount(proposed_block_height) < MIN_VALIDATOR_COUNT)
    return false;
```

MichaelKim20 replied:
I understood your suggestion accurately. Thank you very much for your good opinion.

MichaelKim20 replied:
I realized late that `ConsensusData.enrollments` and `ValidatorSet` cannot have the same data. @AndrejMitrovic , I'm sorry. It's all fixed.

AndrejMitrovic commented:
I will review it again soon.

linked0 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -319,6 +319,33 @@ public class ValidatorSet
+        ulong validator_count = 0;
+        try
+        {
+            ulong height = (block_height >= Enrollment.ValidatorCycle) ?
```
The `height` could be better with the name of `start_height` because the `enrolled_height` column of the `validator_set` table is the height when the enrollment information is stored in a block and the `start_height` is actually the `enrolled_height` plus 1.

MichaelKim20 replied:
I thought after watching `ClearExpiredValidators`.
Please explain in detail what the problem is.

MichaelKim20 replied:
ClearExpiredValidators : 
enrolled_height <= block_height - Enrollment.ValidatorCycle
getValidatorCount:
enrolled_height >= block_height - Enrollment.ValidatorCycle + 1

AndrejMitrovic replied:
I think Michael's code is correct. Consider:

- Validator V is enrolled at block height: `#1`
- V validates and signs blocks `[#2 .. #1009]` inclusively.
- The Ledger is currently at block `#1008`.
- We are nominating block `#1009`.
- We need to ensure there are still active validators which can nominate and sign blocks `#1010+`.
- We call `getValidatorCount(1009)`.
- V should not be counted in this call. Because after block `#1009` is externalized V cannot validate or sign blocks #1010+. Its range was `[#2 .. #1009]`.

So: `getValidatorCount(1009)`:
- => 1009 - 1008 + 1 = 2
- => `SELECT count(*) FROM validator_set WHERE enrolled_height >= 2` will return 0, which is correct. This validator will not be active after block 1009 is externalized.

linked0 replied:
I thought the function finds the  count of validators for the the height that is passed in. I mean, I thought the function finds the count of validators for nominating block you described. I guess I misunderstand the meaning of the block height.

AndrejMitrovic replied:
Right, it's a misnomer. It's not about who can validate a given block, but rather who will be active at that block height so they can validate height + 1 and so on.

Maybe we can find a better name for the parameter.. or document the function better.

linked0 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -319,6 +319,33 @@ public class ValidatorSet
+            ulong height = (block_height >= Enrollment.ValidatorCycle) ?
+                block_height - Enrollment.ValidatorCycle + 1 : 0;
+            () @trusted {
+                validator_count = this.db.execute("SELECT count(*) FROM validator_set WHERE enrolled_height >= ?", height).oneValue!ulong;
```
I think the code below is right because the `height` value that you have calculated is actually the `enrolled_height` plus 1.

```
validator_count = this.db.execute("SELECT count(*) FROM validator_set WHERE enrolled_height >= ?", height-1).oneValue!ulong;
```

MichaelKim20 replied:
I thought after watching ClearExpiredValidators.
Please explain in detail what the problem is.
ClearExpiredValidators :
enrolled_height <= block_height - Enrollment.ValidatorCycle
getValidatorCount:
enrolled_height >= block_height - Enrollment.ValidatorCycle + 1

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1091,3 +1105,115 @@ unittest
+
+    block_height = 1010; // valid block height : [3 .. 1011]
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 2);
```
I think this should be 3 if because the first enrollment has the `enrolled_height` of 2 so the start height as a validator is 3 as you described. I have been thinking the `block_height` you passed in as parameter is the height where the new block is being made.
(이 assert 문에서 체크해야할 값은 3이 되어야한다고 생각합니다. 왜냐하면, 첫번째 enrollment의 enrolled_height는 2이고, validator로서 기능을 시작하는 것은 height가 3이 됩니다. 따라서, 첫번째 enrollment는 3부터 1010까지 validator로서 유효하기 때문에 위 assert 체크값은 3이 되어야 합니다.)

MichaelKim20 replied:
I thought after watching ClearExpiredValidators.
Please explain in detail what the problem is.
ClearExpiredValidators :
enrolled_height <= block_height - Enrollment.ValidatorCycle
getValidatorCount:
enrolled_height >= block_height - Enrollment.ValidatorCycle + 1

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1091,3 +1105,115 @@ unittest
+    assert(man.getValidatorCount(block_height) + enrollments.length == 2);
+
+    man.clearExpiredValidators(block_height);
+    foreach (ref enrollment; enrollments)
```
The `enrollments` in all the `foreach` statement always has just one element, so why don't you use just the variable of `Enrollment` type and not to use `foreach` statement?

MichaelKim20 replied:
I thought after watching ClearExpiredValidators.
Please explain in detail what the problem is.
ClearExpiredValidators :
enrolled_height <= block_height - Enrollment.ValidatorCycle
getValidatorCount:
enrolled_height >= block_height - Enrollment.ValidatorCycle + 1

MichaelKim20 replied:
I changed it.
Thank you for your good suggestion.

linked0 reviewed (commented):
Good job but there are some codes to be fixed.
This comment should have posted earlier though.

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -319,6 +319,33 @@ public class ValidatorSet
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error {}", ex);
```
We will need to handle this better in the future. But it might be left for a future PR.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1107,103 @@ unittest
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 2);
+
+    block_height = 1011; // valid block height : [4 .. 1012]
```
```suggestion
    block_height = 1011; // valid block height : [4 .. 1011]
```

Because at block 1012 there will be zero validators.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1107,103 @@ unittest
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 1);
+
+    block_height = 1012; // valid block height : [5 .. 1013]
```
```suggestion
    block_height = 1012; // valid block height : [4 .. 1011]
```

Because at block 1012 there will be zero validators.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1107,103 @@ unittest
+    import agora.common.Amount;
+    import agora.consensus.data.Transaction;
+    import agora.consensus.Genesis;
+    import std.algorithm;
```
Already imported at module scope

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1107,103 @@ unittest
+    assert(man.createEnrollment(utxo_hash1, block_height, enrollment));
+    assert(!man.hasEnrollment(utxo_hash1));
+    assert(man.add(enrollment, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
```
I don't think we should check this. The enrollment pool is an internal detail and we don't really care if it's 1, 10 or 100 here.

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -319,6 +319,33 @@ public class ValidatorSet
+        {
+            log.error("Database operation error {}", ex);
+        }
+        return validator_count;
```
0 is a valid value to return, but how is this function going to be used ?

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1107,103 @@ unittest
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 3);
+
+    block_height = 1009; // valid block height : [2 .. 1009]
```
I can see two options here:
```suggestion
    block_height = 1009; // valid block height : [2, 1009]
```
OR:
```suggestion
    block_height = 1009; // valid block height : [2 .. 1010]
```

I have a slight preference for the second one, but can see how it could be confusing.
But if we are using the D syntax for array subscription, we have to make it non-inclusive at the end.
In mathematical notation, we would use `[)`, not `[]`, for this.

Geod24 reviewed (commented):
Broad LGTM.
Remove the tests which touch the enrollment pool (the feature is about the *validator set*), and a few other comments, and we're good to go.

MichaelKim20 commented:
@Geod24 Ready for review

MichaelKim20 committed:
Add the ability to obtain the number of valid validators

It will be used as a basis for determining the validity of block content in the consensus protocol

MichaelKim20 commented:
@Geod24 Ready for review

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1093,3 +1116,99 @@ unittest
+    man.clearExpiredValidators(block_height);
+    man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    block_height = 5;    // valid block height : 0 <= H < 1008
```
Nice, that's even better than what I suggested!
